File: windows-app/Plantoir/Views/PublishingChoiceView.cs
using System;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Automation;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Media;
using Plantoir.Core.Models;

namespace Plantoir.Views;

/// <summary>
/// The publishing choice — Netlify, Cloudflare Pages, or a folder on this PC —
/// used by BOTH Course Settings and the new-course wizard, so the two offer
/// exactly the same behaviour and wording (rows 101–102; mirrors the mac's
/// PublishingChoiceView). Validation is live: the problem line follows every
/// keystroke and every choice, and callers gate Save/Create on
/// <see cref="Problem"/> so a publish can never discover a bad setting after
/// the fact.
///
/// The Cloudflare account ID is asked for here rather than left to the
/// launcher's console prompt, because the app publishes with nothing attached
/// to answer a prompt — and a token scoped to Pages cannot look its own
/// account up (checked against a real token).
/// </summary>
public sealed class PublishingChoiceView
{
    public StackPanel Root { get; }

    /// <summary>Raised after every change the teacher makes here.</summary>
    public event Action? Changed;

    private const int NetlifyIndex = 0;
    private const int CloudflareIndex = 1;
    private const int FolderIndex = 2;

    private readonly Func<string> _getTarget;
    private readonly Func<string> _getPath;
    private readonly Func<string> _getAccount;
    private readonly Action<string> _setAccount;
    private readonly Func<IReadOnlyList<CourseConfiguration.AdditionalDeployTarget>> _getAdditional;
    private readonly Action<IReadOnlyList<CourseConfiguration.AdditionalDeployTarget>> _setAdditional;
    private readonly StackPanel _folderArea;
    private readonly StackPanel _cloudflareArea;
    private readonly StackPanel _additionalArea;
    private readonly TextBox _pathBox;
    private readonly TextBox _accountBox;
    private readonly TextBlock _problemText;
    private readonly TextBlock _caption;
    private readonly TextBlock _cloudflareProblemText;
    private readonly TextBlock _cloudflareCaption;
    private readonly TextBlock _cloudflareSizeNote;
    private readonly Window _pickerOwner;
    private bool _updatingFromModel;

    /// <summary>
    /// The Account ID field shown inside the "Also publish to" row when
    /// Cloudflare is the ADDITIONAL destination and the primary is something
    /// else — the primary's own Cloudflare block (<see cref="_accountBox"/>)
    /// is collapsed in that case, so without this field there would be
    /// nowhere on screen to satisfy <see cref="Problem"/>'s requirement for
    /// one, and Save would stay disabled with no way to fix it (found by
    /// Russell 2026-08-22: toggling on "Also deploy to Cloudflare" with
    /// Netlify as the primary target). Kept in sync with <see cref="_accountBox"/>
    /// by <see cref="SyncAccountBoxes"/> since both write the same shared
    /// value — mirrors the mac's <c>CloudflareDetailFields</c> reused inside
    /// its additional-target row.
    /// </summary>
    private TextBox? _additionalCloudflareAccountBox;
    private TextBlock? _additionalCloudflareProblemText;

    /// <summary>
    /// What is wrong with the current choice, or null when nothing is —
    /// always null in Netlify mode, which needs no settings here. Checks
    /// every ADDITIONAL destination too, for the same reason the primary's
    /// own check exists: a bad folder or missing Cloudflare Account ID
    /// blocks Save/Create, rather than failing silently the first time a
    /// deploy actually reaches it. Mirrors the mac's PublishingChoiceView.
    /// </summary>
    public string? Problem
    {
        get
        {
            string? primaryProblem = _getTarget() switch
            {
                "local_folder" => CourseConfiguration.DeployFolderProblem(_getPath()),
                "cloudflare_pages" => CourseConfiguration.CloudflareAccountProblem(_getAccount()),
                _ => null,
            };
            if (primaryProblem is not null) return primaryProblem;
            foreach (var target in _getAdditional())
            {
                string? problem = target.Type switch
                {
                    "local_folder" => CourseConfiguration.DeployFolderProblem(target.Path),
                    "cloudflare_pages" => CourseConfiguration.CloudflareAccountProblem(_getAccount()),
                    _ => null,
                };
                if (problem is not null) return problem;
            }
            return null;
        }
    }

    public PublishingChoiceView(Window pickerOwner,
                                Func<string> getTarget, Action<string> setTarget,
                                Func<string> getPath, Action<string> setPath,
                                Func<string> getAccount, Action<string> setAccount,
                                Func<IReadOnlyList<CourseConfiguration.AdditionalDeployTarget>> getAdditional,
                                Action<IReadOnlyList<CourseConfiguration.AdditionalDeployTarget>> setAdditional)
    {
        _pickerOwner = pickerOwner;
        _getTarget = getTarget;
        _getPath = getPath;
        _getAccount = getAccount;
        _setAccount = setAccount;
        _getAdditional = getAdditional;
        _setAdditional = setAdditional;

        Root = new StackPanel { Spacing = 6 };

        var targetBox = new ComboBox { MinWidth = 300 };
        targetBox.Items.Add("Netlify");
        targetBox.Items.Add("Cloudflare Pages");
        targetBox.Items.Add("A folder on this PC");
        targetBox.SelectedIndex = getTarget() switch
        {
            "local_folder" => FolderIndex,
            "cloudflare_pages" => CloudflareIndex,
            _ => NetlifyIndex,
        };
        AutomationProperties.SetAutomationId(targetBox, "deployTargetPicker");
        Root.Children.Add(FormBuilders.LabeledRow("Deploy to", targetBox));

        // ---- Folder ----
        _folderArea = new StackPanel { Spacing = 4 };
        var pathRow = new Grid { ColumnSpacing = 8 };
        pathRow.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
        pathRow.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
        _pathBox = new TextBox { Text = getPath(), PlaceholderText = "Folder" };
        AutomationProperties.SetAutomationId(_pathBox, "deployFolderField");
        pathRow.Children.Add(_pathBox);
        var chooseButton = new Button { Content = "Choose…" };
        AutomationProperties.SetAutomationId(chooseButton, "deployFolderChooseButton");
        Grid.SetColumn(chooseButton, 1);
        pathRow.Children.Add(chooseButton);
        _folderArea.Children.Add(pathRow);

        _problemText = CautionLine("deployFolderProblem");
        _folderArea.Children.Add(_problemText);
        _caption = FormBuilders.ExampleCaption(
            "Each section deploys into its own subfolder here — section1, section2 — and only changed files are copied. Upload the folder to your web host however you prefer (e.g. over SFTP). Netlify isn’t involved.");
        _folderArea.Children.Add(_caption);
        Root.Children.Add(_folderArea);

        // ---- Cloudflare ----
        _cloudflareArea = new StackPanel { Spacing = 4 };
        _accountBox = new TextBox { Text = getAccount(), PlaceholderText = "Account ID" };
        AutomationProperties.SetAutomationId(_accountBox, "cloudflareAccountField");
        _cloudflareArea.Children.Add(FormBuilders.LabeledRow("Cloudflare Account ID", _accountBox));

        var helpButton = new HyperlinkButton
        {
            Content = "Where do I find this?",
            FontSize = 12,
            Padding = new Thickness(0),
            Margin = new Thickness(0, 2, 0, 4),
        };
        helpButton.Click += async (_, _) =>
        {
            Plantoir.Core.Scripting.ActivityTrail.Note(
                Plantoir.Core.Scripting.ActivityTrail.Event.AskedForACredential,
                "opened instructions for Cloudflare Account ID");
            var help = Plantoir.Core.Scripting.CredentialRequests.CloudflareAccountIDHelp;
            var panel = CredentialRequestPanel.Build(help);

            var dialog = new ContentDialog
            {
                Title = help.Title,
                Content = panel,
                CloseButtonText = "Done",
                DefaultButton = ContentDialogButton.Close,
                XamlRoot = _pickerOwner.Content?.XamlRoot
            };
            await dialog.ShowAsync();
        };
        _cloudflareArea.Children.Add(helpButton);

        _cloudflareProblemText = CautionLine("cloudflareAccountProblem");
        _cloudflareArea.Children.Add(_cloudflareProblemText);
        _cloudflareCaption = FormBuilders.ExampleCaption(
            "Each section gets its own free site at yourproject.pages.dev. Sign in at dash.cloudflare.com and open Workers and Pages — the Account ID is shown on the right, and it’s also the long code in the address bar. You only enter it once. The first deploy asks for an API token with the Cloudflare Pages permission.");
        _cloudflareArea.Children.Add(_cloudflareCaption);

        // Stays on screen the whole time Cloudflare is chosen — it is a
        // standing fact about the destination, not a mistake to correct, so
        // it is worded as a heads-up and never hides the way the problem
        // line does.
        _cloudflareSizeNote = CautionLine("cloudflareSizeNote");
        // The area's own visibility already hides this with everything else
        // when another destination is chosen.
        _cloudflareSizeNote.Visibility = Visibility.Visible;
        _cloudflareSizeNote.Text =
            "One thing to know: Cloudflare won’t accept any single file larger than 25 MB. " +
            "Documents, images, and slide decks are almost always comfortably under that — a long video usually isn’t. " +
            "Most teachers embed video from YouTube or Vimeo rather than uploading it, which avoids the limit entirely.";
        _cloudflareArea.Children.Add(_cloudflareSizeNote);
        Root.Children.Add(_cloudflareArea);

        // ---- Also publish to, for redundancy — real redundancy needs a
        // second copy of the site ALREADY live, not a scramble to
        // reconfigure a new destination after the fact. See
        // documentation/07-deployment.md.
        _additionalArea = new StackPanel { Spacing = 6, Margin = new Thickness(0, 12, 0, 0) };
        Root.Children.Add(_additionalArea);

        targetBox.SelectionChanged += (_, _) =>
        {
            if (_updatingFromModel) return;
            setTarget(targetBox.SelectedIndex switch
            {
                FolderIndex => "local_folder",
                CloudflareIndex => "cloudflare_pages",
                _ => "netlify",
            });
            RefreshAreas();
            RebuildAdditionalArea();
            Changed?.Invoke();
        };
        _pathBox.TextChanged += (_, _) =>
        {
            if (_updatingFromModel) return;
            setPath(_pathBox.Text);
            RefreshAreas();
            Changed?.Invoke();
        };
        _accountBox.TextChanged += (_, _) =>
        {
            if (_updatingFromModel) return;
            setAccount(_accountBox.Text.Trim());
            SyncAccountBoxes(_accountBox);
            RefreshAreas();
            Changed?.Invoke();
        };
        chooseButton.Click += async (_, _) =>
        {
            var picker = new Windows.Storage.Pickers.FolderPicker();
            WinRT.Interop.InitializeWithWindow.Initialize(picker,
                WinRT.Interop.WindowNative.GetWindowHandle(_pickerOwner));
            picker.FileTypeFilter.Add("*");
            var folder = await picker.PickSingleFolderAsync();
            if (folder is null) return;
            _updatingFromModel = true;
            _pathBox.Text = folder.Path;
            _updatingFromModel = false;
            setPath(folder.Path);
            RefreshAreas();   // validation runs on the choice instantly
            Changed?.Invoke();
        };

        RefreshAreas();
        RebuildAdditionalArea();
    }

    private static string FriendlyName(string type) => type switch
    {
        "local_folder" => "A folder on this PC",
        "cloudflare_pages" => "Cloudflare Pages",
        _ => "Netlify",
    };

    /// <summary>
    /// Rebuilt (not just re-shown) whenever the primary changes, because
    /// the SET of available additional types changes with it — "one of
    /// each type, and never the primary twice" (row 304).
    /// </summary>
    private void RebuildAdditionalArea()
    {
        _additionalArea.Children.Clear();
        _additionalCloudflareAccountBox = null;
        _additionalCloudflareProblemText = null;
        var availableTypes = CourseConfiguration.AvailableAdditionalDeployTargetTypes(_getTarget());
        if (availableTypes.Count == 0) return;

        _additionalArea.Children.Add(new TextBlock
        {
            Text = "Also publish to, for redundancy",
            FontWeight = Microsoft.UI.Text.FontWeights.SemiBold,
        });

        foreach (string type in availableTypes)
        {
            var toggle = new ToggleSwitch
            {
                OnContent = "", OffContent = "",
                IsOn = CourseConfiguration.HasAdditionalDeployTarget(_getAdditional(), type),
            };
            AutomationProperties.SetAutomationId(toggle, $"additionalDeployTarget-{type}");
            var row = FormBuilders.LabeledRow(FriendlyName(type), toggle);
            _additionalArea.Children.Add(row);

            StackPanel? detailArea = null;
            TextBox? folderBox = null;
            TextBlock? folderProblemText = null;

            if (type == "local_folder")
            {
                detailArea = new StackPanel { Spacing = 4, Margin = new Thickness(0, 0, 0, 4) };
                var pathRow = new Grid { ColumnSpacing = 8 };
                pathRow.ColumnDefinitions.Add(new ColumnDefinition { Width = new GridLength(1, GridUnitType.Star) });
                pathRow.ColumnDefinitions.Add(new ColumnDefinition { Width = GridLength.Auto });
                folderBox = new TextBox
                {
                    Text = CourseConfiguration.AdditionalDeployTargetPath(_getAdditional(), type),
                    PlaceholderText = "Folder",
                };
                AutomationProperties.SetAutomationId(folderBox, $"additionalDeployFolderField-{type}");
                pathRow.Children.Add(folderBox);
                var chooseAdditional = new Button { Content = "Choose…" };
                Grid.SetColumn(chooseAdditional, 1);
                pathRow.Children.Add(chooseAdditional);
                detailArea.Children.Add(pathRow);

                folderProblemText = CautionLine($"additionalDeployFolderProblem-{type}");
                detailArea.Children.Add(folderProblemText);

                chooseAdditional.Click += async (_, _) =>
                {
                    var picker = new Windows.Storage.Pickers.FolderPicker();
                    WinRT.Interop.InitializeWithWindow.Initialize(picker,
                        WinRT.Interop.WindowNative.GetWindowHandle(_pickerOwner));
                    picker.FileTypeFilter.Add("*");
                    var folder = await picker.PickSingleFolderAsync();
                    if (folder is null) return;
                    folderBox.Text = folder.Path;
                    _setAdditional(CourseConfiguration.SettingAdditionalDeployTargetPath(_getAdditional(), folder.Path, type));
                    RefreshAdditionalProblem(folderProblemText, type);
                    Changed?.Invoke();
                };
                folderBox.TextChanged += (_, _) =>
                {
                    _setAdditional(CourseConfiguration.SettingAdditionalDeployTargetPath(_getAdditional(), folderBox.Text, type));
                    RefreshAdditionalProblem(folderProblemText, type);
                    Changed?.Invoke();
                };
                RefreshAdditionalProblem(folderProblemText, type);
                _additionalArea.Children.Add(detailArea);
            }
            else if (type == "cloudflare_pages")
            {
                // Cloudflare's account ID is per-teacher, in app settings —
                // NOT stored per destination — so this reuses the exact same
                // value the primary picker's own block asks for. But that
                // block is only ON SCREEN when the primary destination IS
                // Cloudflare; when it isn't (e.g. primary is Netlify), a
                // note pointing at it points at nothing the teacher can see,
                // and Problem still requires an account ID before Save can
                // enable — so a real field belongs here too, kept in sync
                // with the primary one by SyncAccountBoxes.
                detailArea = new StackPanel { Spacing = 4, Margin = new Thickness(0, 0, 0, 4) };
                var accountBox = new TextBox
                {
                    Text = _getAccount(),
                    PlaceholderText = "Account ID",
                };
                AutomationProperties.SetAutomationId(accountBox, "additionalCloudflareAccountField");
                detailArea.Children.Add(FormBuilders.LabeledRow("Cloudflare Account ID", accountBox));

                var problemText = CautionLine("additionalCloudflareAccountProblem");
                detailArea.Children.Add(problemText);
                _additionalCloudflareAccountBox = accountBox;
                _additionalCloudflareProblemText = problemText;
                RefreshAdditionalCloudflareProblem();

                accountBox.TextChanged += (_, _) =>
                {
                    if (_updatingFromModel) return;
                    _setAccount(accountBox.Text.Trim());
                    SyncAccountBoxes(accountBox);
                    RefreshAreas();
                    Changed?.Invoke();
                };
                _additionalArea.Children.Add(detailArea);
            }

            string capturedType = type;
            StackPanel? capturedDetailArea = detailArea;
            toggle.Toggled += (_, _) =>
            {
                _setAdditional(CourseConfiguration.SettingAdditionalDeployTarget(_getAdditional(), toggle.IsOn, capturedType));
                if (capturedDetailArea is not null)
                    capturedDetailArea.Visibility = toggle.IsOn ? Visibility.Visible : Visibility.Collapsed;
                Changed?.Invoke();
            };
            if (detailArea is not null)
                detailArea.Visibility = toggle.IsOn ? Visibility.Visible : Visibility.Collapsed;
        }
    }

    private void RefreshAdditionalProblem(TextBlock problemText, string type)
    {
        string path = CourseConfiguration.AdditionalDeployTargetPath(_getAdditional(), type);
        string? problem = CourseConfiguration.DeployFolderProblem(path);
        problemText.Text = problem ?? "";
        problemText.Visibility = problem is null ? Visibility.Collapsed : Visibility.Visible;
    }

    /// <summary>
    /// The Account ID is one shared value, but it can be typed into either
    /// of two text boxes (the primary Cloudflare block, or the additional-
    /// target row's own field) depending on which is on screen — keeps
    /// whichever box the teacher did NOT just type into showing the same
    /// text, so switching the primary destination later shows what was
    /// actually saved rather than a stale value from construction time.
    /// </summary>
    private void SyncAccountBoxes(TextBox editedBox)
    {
        _updatingFromModel = true;
        string value = _getAccount();
        if (!ReferenceEquals(editedBox, _accountBox)) _accountBox.Text = value;
        if (_additionalCloudflareAccountBox is not null && !ReferenceEquals(editedBox, _additionalCloudflareAccountBox))
            _additionalCloudflareAccountBox.Text = value;
        _updatingFromModel = false;
    }

    private void RefreshAdditionalCloudflareProblem()
    {
        if (_additionalCloudflareProblemText is null) return;
        string? problem = CourseConfiguration.CloudflareAccountProblem(_getAccount());
        _additionalCloudflareProblemText.Text = problem ?? "";
        _additionalCloudflareProblemText.Visibility = problem is null ? Visibility.Collapsed : Visibility.Visible;
    }

    private static TextBlock CautionLine(string automationId)
    {
        var line = new TextBlock
        {
            FontSize = 12,
            TextWrapping = TextWrapping.Wrap,
            Foreground = (Brush)Application.Current.Resources["SystemFillColorCautionBrush"],
            Visibility = Visibility.Collapsed,
        };
        AutomationProperties.SetAutomationId(line, automationId);
        return line;
    }

    private void RefreshAreas()
    {
        string target = _getTarget();
        bool folderMode = target == "local_folder";
        bool cloudflareMode = target == "cloudflare_pages";
        _folderArea.Visibility = folderMode ? Visibility.Visible : Visibility.Collapsed;
        _cloudflareArea.Visibility = cloudflareMode ? Visibility.Visible : Visibility.Collapsed;

        // Only the visible area's problem is shown; Problem itself already
        // reports on whichever mode is active.
        string? problem = Problem;
        ShowProblem(_problemText, _caption, folderMode ? problem : null);
        ShowProblem(_cloudflareProblemText, _cloudflareCaption, cloudflareMode ? problem : null);
        RefreshAdditionalCloudflareProblem();
    }

    private static void ShowProblem(TextBlock line, TextBlock caption, string? problem)
    {
        line.Text = problem ?? "";
        line.Visibility = problem is null ? Visibility.Collapsed : Visibility.Visible;
        caption.Visibility = problem is null ? Visibility.Visible : Visibility.Collapsed;
    }
}

File: mac-app/QuartzTeachers/Views/CourseSettings/PublishingChoiceView.swift
import AppKit
import SwiftUI

/// The deploy destination(s) — Netlify, Cloudflare Pages, or a folder on
/// this Mac — used by both Course Settings and the new-course wizard, so
/// the two offer exactly the same behaviour and wording.
///
/// One is always the PRIMARY (the picker at the top); a teacher may also
/// switch on any of the other known types as ADDITIONAL destinations, for
/// redundancy against one host having a bad day. Deploying publishes to
/// every configured destination in one action — this view only decides
/// which destinations are configured, not how deploying uses them.
struct PublishingChoiceView: View {

    // MARK: - Stored properties

    @Binding var deployTarget: String
    @Binding var deployFolderPath: String

    /// The teacher's Cloudflare Account ID. It lives in app settings, not
    /// in this course's settings — it identifies the person, not the class
    /// — so it is entered once and every Cloudflare course uses it,
    /// whether Cloudflare is the primary destination or an additional one.
    @Binding var cloudflareAccountID: String

    /// Every OTHER destination this course also publishes to, beyond the
    /// primary above — see `CourseConfiguration.additionalDeployTargets`.
    @Binding var additionalDeployTargets: [CourseConfiguration.AdditionalDeployTarget]

    /// True while the "Where do I find this?" instructions are open.
    @State var isShowingAccountHelp: Bool = false

    // MARK: - Computed properties

    /// What is wrong with the chosen folder, or nil when nothing is.
    /// Checks the PRIMARY folder only — an additional folder target's
    /// problem is shown inline beside its own field instead.
    var folderProblem: String? {
        if deployTarget != "local_folder" {
            return nil
        }
        return CourseConfiguration.deployFolderProblem(forPath: deployFolderPath)
    }

    /// True when Cloudflare is configured as an ADDITIONAL destination.
    /// Never true at the same time as `deployTarget == "cloudflare_pages"`
    /// — the picker's own `onChange` below keeps the two from disagreeing.
    var hasAdditionalCloudflare: Bool {
        for target in additionalDeployTargets where target.type == "cloudflare_pages" {
            return true
        }
        return false
    }

    /// True when a local folder is configured as an ADDITIONAL destination.
    var hasAdditionalLocalFolder: Bool {
        for target in additionalDeployTargets where target.type == "local_folder" {
            return true
        }
        return false
    }

    /// What is wrong with the Cloudflare Account ID, or nil when nothing
    /// is. The same ID serves Cloudflare whether it is primary or
    /// additional — the two are mutually exclusive, so this applies to
    /// whichever one is actually configured.
    var accountProblem: String? {
        if deployTarget != "cloudflare_pages" && !hasAdditionalCloudflare {
            return nil
        }
        return CourseConfiguration.cloudflareAccountProblem(forID: cloudflareAccountID)
    }

    /// The other destination types this course could ALSO publish to —
    /// every known type except whichever one is currently primary.
    var availableAdditionalDeployTargetTypes: [String] {
        var result: [String] = []
        for type in CourseConfiguration.knownDeployTargetTypes where type != deployTarget {
            result.append(type)
        }
        return result
    }

    /// A live folder-path editor for the additional local-folder target.
    var additionalFolderPathBinding: Binding<String> {
        return Binding(
            get: {
                for target in additionalDeployTargets where target.type == "local_folder" {
                    return target.path
                }
                return ""
            },
            set: { newPath in
                var targets: [CourseConfiguration.AdditionalDeployTarget] = []
                for target in additionalDeployTargets {
                    if target.type == "local_folder" {
                        targets.append(CourseConfiguration.AdditionalDeployTarget(type: "local_folder", path: newPath))
                    } else {
                        targets.append(target)
                    }
                }
                additionalDeployTargets = targets
            }
        )
    }

    /// What is wrong with the additional local-folder target's path, or
    /// nil when nothing is (or it isn't configured).
    var additionalFolderProblem: String? {
        if !hasAdditionalLocalFolder {
            return nil
        }
        return CourseConfiguration.deployFolderProblem(forPath: additionalFolderPathBinding.wrappedValue)
    }

    /// Anything that must be put right before this course can be saved.
    var problem: String? {
        if let folderProblem {
            return folderProblem
        }
        if let accountProblem {
            return accountProblem
        }
        return additionalFolderProblem
    }

    // MARK: - Body

    var body: some View {
        Picker("Deploy to", selection: $deployTarget) {
            ForEach(CourseConfiguration.knownDeployTargetTypes, id: \.self) { type in
                Text(PublishingChoiceView.label(forDeployTargetType: type)).tag(type)
            }
        }
        .accessibilityIdentifier("deployTargetPicker")
        .onChange(of: deployTarget) { _, newValue in
            // Keeps the WIZARD's plain @State consistent live on screen —
            // it has no CourseConfiguration to route through until the
            // course is actually created, so this view enforces the same
            // rule `CourseConfiguration.deployTarget`'s setter enforces
            // for Course Settings, which binds straight to the model.
            additionalDeployTargets = CourseConfiguration.pruningAdditionalTargets(
                additionalDeployTargets, ofType: newValue
            )
        }

        if deployTarget == "cloudflare_pages" {
            CloudflareDetailFields(
                cloudflareAccountID: $cloudflareAccountID,
                isShowingAccountHelp: $isShowingAccountHelp,
                accountProblem: accountProblem
            )
        }

        if deployTarget == "local_folder" {
            LocalFolderDetailFields(
                path: $deployFolderPath,
                problem: folderProblem,
                chooseIdentifier: "deployFolderChooseButton",
                fieldIdentifier: "deployFolderField",
                problemIdentifier: "deployFolderProblem"
            )
        }

        if !availableAdditionalDeployTargetTypes.isEmpty {
            VStack(alignment: .leading, spacing: 8) {
                Text("Also publish to, for redundancy")
                    .font(.subheadline)
                    .foregroundStyle(.secondary)
                ExampleCaption("Deploying publishes to every destination switched on here, one after another. If one host is down or having trouble, the others still go out. Most teachers leave this off — it is here for anyone who wants a second copy live.")

                ForEach(availableAdditionalDeployTargetTypes, id: \.self) { type in
                    additionalTargetRow(forType: type)
                }
            }
        }
    }

    /// One row of the additional-targets list: its toggle, and — only
    /// while switched on — the same detail fields the primary destination
    /// would show for that type.
    @ViewBuilder
    func additionalTargetRow(forType type: String) -> some View {
        Toggle(
            PublishingChoiceView.label(forDeployTargetType: type),
            isOn: toggleBinding(forType: type)
        )
        .accessibilityIdentifier("additionalDeployTargetToggle-\(type)")

        if type == "cloudflare_pages" && hasAdditionalCloudflare {
            CloudflareDetailFields(
                cloudflareAccountID: $cloudflareAccountID,
                isShowingAccountHelp: $isShowingAccountHelp,
                accountProblem: accountProblem
            )
            .padding(.leading, 20)
        }

        if type == "local_folder" && hasAdditionalLocalFolder {
            LocalFolderDetailFields(
                path: additionalFolderPathBinding,
                problem: additionalFolderProblem,
                chooseIdentifier: "additionalDeployFolderChooseButton",
                fieldIdentifier: "additionalDeployFolderField",
                problemIdentifier: "additionalDeployFolderProblem"
            )
            .padding(.leading, 20)
        }
    }

    // MARK: - Functions

    /// The label a teacher sees for a destination type — shared by the
    /// primary picker and the additional-targets list, so the two never
    /// describe the same destination in different words.
    static func label(forDeployTargetType type: String) -> String {
        switch type {
        case "netlify":
            return "Netlify"
        case "cloudflare_pages":
            return "Cloudflare Pages"
        case "local_folder":
            return "A folder on this Mac"
        default:
            return type
        }
    }

    /// A live on/off switch for one additional-target type, backed by the
    /// array binding above.
    func toggleBinding(forType type: String) -> Binding<Bool> {
        return Binding(
            get: {
                for target in additionalDeployTargets where target.type == type {
                    return true
                }
                return false
            },
            set: { isOn in
                var targets: [CourseConfiguration.AdditionalDeployTarget] = []
                for target in additionalDeployTargets where target.type != type {
                    targets.append(target)
                }
                if isOn {
                    targets.append(CourseConfiguration.AdditionalDeployTarget(type: type, path: ""))
                }
                additionalDeployTargets = targets
            }
        )
    }
}

/// The Cloudflare Account ID field, its help button, its own validation
/// message, and the fixed 25 MB caption — shared between the primary
/// destination's detail block and an additional Cloudflare target's,
/// which are never both on screen at once (a type is either primary or
/// additional, never both).
private struct CloudflareDetailFields: View {

    // MARK: - Stored properties

    @Binding var cloudflareAccountID: String
    @Binding var isShowingAccountHelp: Bool
    let accountProblem: String?

    // MARK: - Body

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            TextField("Cloudflare Account ID", text: $cloudflareAccountID)
                .textFieldStyle(.roundedBorder)
                .accessibilityIdentifier("cloudflareAccountField")

            // The grey caption that used to sit here repeated the
            // dashboard directions in four lines of small text under a
            // field that wants one code, whether or not anybody had a
            // question. It is a dialog now, on request. The Safari mark
            // matches the dialogs, so a link that leaves the app looks
            // the same everywhere.
            Button {
                showAccountHelp()
            } label: {
                Label("Where do I find this?", systemImage: "safari")
            }
            .buttonStyle(.link)
            .accessibilityIdentifier("cloudflareAccountHelpButton")

            // What is WRONG with what was typed. Not commentary: this
            // is why the course will not save.
            if let accountProblem {
                Text(accountProblem)
                    .font(.caption)
                    .foregroundStyle(.orange)
                    .accessibilityIdentifier("cloudflareAccountProblem")
            }

            // Not a warning that comes and goes: a fact about this
            // destination that never hides. It is the one real
            // functional difference between the destinations, so a
            // teacher should meet it while choosing rather than when
            // a deploy fails halfway through an upload.
            Text("One thing to know: Cloudflare won’t accept any single file larger than 25 MB. Documents, images, and slide decks are almost always comfortably under that — a long video usually isn’t. Most teachers embed video from YouTube or Vimeo rather than uploading it, which avoids the limit entirely.")
                .font(.caption)
                .foregroundStyle(.orange)
                .accessibilityIdentifier("cloudflareSizeNote")
        }
        // The same dialog the launcher shows when it has to ask for an
        // Account ID mid-publish, opened here on purpose instead. What
        // is typed into it lands in the field above, so a teacher who
        // has just fetched the code does not have to close this and
        // find the field again.
        .sheet(isPresented: $isShowingAccountHelp) {
            CredentialRequestSheet(
                request: CredentialRequest.cloudflareAccountIDHelp,
                initialAnswer: cloudflareAccountID,
                confirmTitle: "Use this ID",
                onSend: { typed in
                    cloudflareAccountID = typed
                    isShowingAccountHelp = false
                },
                onCancel: {
                    isShowingAccountHelp = false
                }
            )
        }
    }

    // MARK: - Functions

    /// Opens the instructions for finding an Account ID.
    ///
    /// Recorded on the trail: a teacher who had to go looking for this is
    /// the same teacher whose first publish is about to fail on it, and the
    /// line says which credential they were looking for. What they type is
    /// never recorded.
    func showAccountHelp() {
        ActivityTrail.note(.askedForACredential, "opened the instructions for finding a Cloudflare Account ID")
        isShowingAccountHelp = true
    }
}

/// The folder path field, its Choose… button, and its own validation
/// message — shared between the primary destination's detail block and
/// an additional local-folder target's.
private struct LocalFolderDetailFields: View {

    // MARK: - Stored properties

    @Binding var path: String
    let problem: String?
    let chooseIdentifier: String
    let fieldIdentifier: String
    let problemIdentifier: String

    // MARK: - Body

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            HStack {
                TextField("Folder", text: $path)
                    .textFieldStyle(.roundedBorder)
                    .accessibilityIdentifier(fieldIdentifier)
                Button("Choose…") {
                    chooseDeployFolder()
                }
                .accessibilityIdentifier(chooseIdentifier)
            }
            if let problem {
                // The same orange every other inline warning wears.
                Text(problem)
                    .font(.caption)
                    .foregroundStyle(.orange)
                    .accessibilityIdentifier(problemIdentifier)
            } else {
                ExampleCaption("Each section deploys into its own subfolder here — section1, section2 — and only changed files are copied. Upload the folder to your web host however you prefer (e.g. over SFTP). Netlify isn’t involved.")
            }
        }
    }

    // MARK: - Functions

    /// The standard folder chooser, writing straight into the setting.
    func chooseDeployFolder() {
        let panel: NSOpenPanel = NSOpenPanel()
        panel.canChooseFiles = false
        panel.canChooseDirectories = true
        panel.canCreateDirectories = true
        panel.prompt = "Choose"
        panel.message = "Choose the folder this course's sections deploy into."
        if panel.runModal() == .OK, let chosen = panel.url {
            path = chosen.path
        }
    }
}
